moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/state_validity_checker.h

```diff
@@ -81,4 +81,60 @@ class StateValidityChecker : public ompl::base::StateValidityChecker
   collision_detection::CollisionRequest collision_request_with_cost_;
   bool verbose_;
 };
+
+/** \brief A StateValidityChecker that can handle states of type `ompl::base::ConstraintStateSpace::StateType`.
+ *
+ * We cannot not just cast the states and pass them to the isValid version of the parent class, because inside the
+ * state-validity checker, the line:
+ *
+ *     planning_context_->getOMPLStateSpace()->copyToRobotState(*robot_state, state);
+ *
+ * requires the state type to be of the constrained state space, while:
+ *
+ *     state->as<ModelBasedStateSpace::StateType>()->isValidityKnown()
+ *
+ * requires accessing the underlying state by calling `getState()` on the constrained state space state.
+ * Therefore this class implements specific versions of the isValid methods.
+ *
+ * We still check the path constraints, because not all states sampled by the constrained state space
+ * satisfy the constraints unfortunately. This is a complicated issue. For more details see:
+ * https://github.com/ros-planning/moveit/issues/2092#issuecomment-669911722.
+ **/
+class ConstrainedPlanningStateValidityChecker : public StateValidityChecker
+{
+public:
+  using StateValidityChecker::isValid;
```

Comment:
Because the `ompl::base::State` pointer passed to the state validity checker has a different type when planning in constrained state space. I need to "unwrap" it to get to the actual state, as is done [here](https://github.com/JeroenDM/moveit/blob/ad0aa16fbb206481fb1394c804176f509b84e2da/moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp#L209) like so:
```C++
  // Unwrap the state from a ConstrainedStateSpace::StateType
  auto state = wrapped_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState();
```

I considered using dynamic casting to check at runtime what kind of pointer it is, but I assumed this would be slower and collision checking is often *the* bottleneck. (I did try this out and it works, but I did not benchmark it...).

Both solutions are equally ugly in my opinion...